### PR TITLE
2.8 into develop

### PR DESCRIPTION
**Pull request by manadart**
Merge from 2.8 to bring forward:
- #11924 from wallyworld/uniter-run-env
- #11925 from wallyworld/k8s-workload-version
- #11922 from achilleasa/2.8-improve-help-text-for-juju-config
- #11921 from manadart/2.8-linklayer-unique-device
- #11916 from achilleasa/2.8-ensure-caas-specific-charm-options-pass-validation-checks
- #11920 from SimonRichardson/github-action-update-mongo-4.4
- #11913 from SimonRichardson/facade-schema-groups
- #11910 from hmlanigan/empty-string-filtering
- #11915 from wallyworld/juju-exec-helptext
- #11902 from ycliuhw/feature/deployment-stragegy
- #11909 from achilleasa/2.8-consistent-support-for-at-syntax-config-inclusion

Some tribulation with conflicts in the following files:
- apiserver/facades/client/application/application.go
- apiserver/facades/client/client/status.go
- cmd/juju/application/deploy.go

### CODE REVIEW

File: apiserver/facades/client/client/status.go [hunk 1]
```diff
@@ -1237,7 +1236,8 @@ func (context *statusContext) processApplication(application *state.Application)
 			sort.Sort(bySinceDescending(versions))
 			processedStatus.WorkloadVersion = versions[0].Message
 		}
-	} else {
+	}
```
Comment (wallyworld) [inline apiserver/facades/client/client/status.go:1239]: This isn't right, the above block to get the workload version is now run for all models.

```
...
	planRequired := metrics != nil && metrics.Plan != nil && metrics.Plan.Required
	if planRequired || len(application.MetricCredentials()) > 0 {
		processedStatus.MeterStatuses = context.processUnitMeterStatuses(units)
	}

	versions := make([]status.StatusInfo, 0, len(units))
	for _, unit := range units {
		workloadVersion, err := context.status.FullUnitWorkloadVersion(unit.Name())
		if err != nil {
			processedStatus.Err = apiservererrors.ServerError(err)
			return processedStatus
		}
		versions = append(versions, workloadVersion)
	}
	if len(versions) > 0 {
		sort.Sort(bySinceDescending(versions))
		processedStatus.WorkloadVersion = versions[0].Message
	}
	if processedStatus.WorkloadVersion == "" && context.model.Type() == state.ModelTypeCAAS {
		// We'll punt on using the docker image name.
		caasModel, err := context.model.CAASModel()
...
```

**Review by wallyworld — CHANGES_REQUESTED** [review]
The changes to facades/client/client/status.go aren't quite right, see comment
The changes to facades/client/application/application.go look ok but best get @achilleasa to double check

**Comment by manadart** [comment]
@wallyworld @hmlanigan See the last commit - I relocated `ReadValue` to the new `utils` sub-package to bring @achilleasa's changes for the "@" syntax over.

**Review by wallyworld — APPROVED** [review]
Thanks! Just a couple of tweaks.

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
$$merge$$

**Comment by manadart** [comment]
I was unable to reproduce this failure:
```
FAIL: firewaller_test.go:1234: InstanceModeSuite.TestRemoteRelationIngressFallbackToPublic
...
[LOG] 0:03.416 INFO test opened port ranges [3306/tcp] on "machine-1"
[LOG] 0:03.490 DEBUG juju.apiserver -> [2] machine-0 87.923215ms {"request-id":54,"response":"'body redacted'"} StringsWatcher["3"].Next
[LOG] 0:03.490 DEBUG juju.apiserver <- [2] machine-0 {"request-id":64,"type":"StringsWatcher","version":1,"id":"3","request":"Next","params":"'params redacted'"}
[LOG] 0:03.490 DEBUG juju.apiserver <- [2] machine-0 {"request-id":65,"type":"RemoteRelations","version":2,"request":"Relations","params":"'params redacted'"}
[LOG] 0:03.491 DEBUG juju.apiserver -> [2] machine-0 560.429µs {"request-id":65,"response":"'body redacted'"} RemoteRelations[""].Relations
[LOG] 0:03.491 DEBUG test relation wordpress:db mysql:server was known but has died or been suspended
[LOG] 0:03.491 DEBUG test forget relation wordpress:db mysql:server
[LOG] 0:03.491 DEBUG test stopped watching "relation-wordpress.db#mysql.server"
[LOG] 0:03.491 DEBUG juju.apiserver <- [2] machine-0 {"request-id":66,"type":"StringsWatcher","version":1,"id":"8","request":"Stop","params":"'params redacted'"}
[LOG] 0:03.491 DEBUG juju.apiserver -> [2] machine-0 396.824892ms {"request-id":30,"response":"'body redacted'"} NotifyWatcher["6"].Next
[LOG] 0:03.491 DEBUG juju.apiserver <- [2] machine-0 {"request-id":67,"type":"NotifyWatcher","version":1,"id":"6","request":"Next","params":"'params redacted'"}
[LOG] 0:03.491 DEBUG juju.apiserver <- [2] machine-0 {"request-id":68,"type":"Firewaller","version":6,"request":"GetExposed","params":"'params redacted'"}
[LOG] 0:03.492 DEBUG juju.apiserver -> [2] machine-0 573.88µs {"request-id":68,"response":"'body redacted'"} Firewaller[""].GetExposed
[LOG] 0:03.497 DEBUG juju.apiserver -> [2] machine-0 6.196612ms {"request-id":66,"response":"'body redacted'"} StringsWatcher["8"].Stop
[LOG] 0:03.497 DEBUG juju.apiserver -> [2] machine-0 86.587273ms {"request-id":61,"error":"watcher has been stopped","error-code":"stopped","response":"'body redacted'"} StringsWatcher["8"].Next
firewaller_test.go:1239:
    s.assertIngressCidrs(c, ingress, []string{"0.0.0.0/0"})
firewaller_test.go:129:
    c.Fatalf("timed out: expected %q; got %q", expected, got)
... Error: timed out: expected []; got ["3306/tcp"]
```
$$merge$$

**Pull request merged**